Clean things up

ISSOtm opened the pull request:
Descriptive PR names are for losers :sunglasses: (more seriously, unless this gets merged bit by bit, this is going to become quite the hodgepodge of changes :D)

ISSOtm committed:
Change `vc_name` definition to be more customizable

And thus more hack-friendly

ISSOtm committed:
Move directory definitions outside of `cia-config.mk`

This file does not get updated by default due to its nature, so we should
move anything the user is not expected to edit out of it

ISSOtm committed:
Use curly braces for Make variable expansions

Just a style choice, but I find it makes parens nesting less difficult to navigate

vulcandth commented:
Just looked over the first few commits; looks great so far.

ISSOtm commented:
Currently working on switching from "multiple-targets: multiple-deps" to implicit rules, and then I'll work on keeping the hack Makefile pristine

ISSOtm commented:
Also, `ctrtool` dumps a f@!#$ckton of things to the console; if there is a flag to silence it, we should enable it by default...

vulcandth commented:
> Also, `ctrtool` dumps a f@!#$ckton of things to the console; if there is a flag to silence it, we should enable it by default...

Yes it does... I don't see a silence option.. only an option to print MORE output.
`-v, --verbose      Give verbose output.`

We might could try redirecting output to /dev/null?

Opened an issue for it:
https://github.com/3DSGuy/Project_CTR/issues/113

ISSOtm commented:
If there's no other way, then yeah, I guess so.

ISSOtm committed:
Use separate rules for building components of the CIA files

The rules currently have one target per `vc_name` entry,
but only generate the files one at a time.
Doing it this way should be more reliable, and I believe more readable
from getting rid of some nested `addprefix`es.

ISSOtm commented:
Done! It builds at least for me, and I'm confident enough in it that I think it should work on other setups :P

ISSOtm committed:
Touch up gitignore

Notably, ignore files only at the root, and ignore the directories
in which the CIAs are extracted

ISSOtm committed:
Simplify variable substitution a bit

ISSOtm committed:
Silence `ctrtool`

It is VERY verbose, holy crap

ISSOtm committed:
Save a subshell and prepare for a better dep chain

ISSOtm commented:
The next thing I would like to fix is relying on the "extract" directory's timestamp; instead, we can either:
- Create a "stamp" file whose sole purpose is to track the extraction time (easier but not the cleanest)
- Explicitly depend on all the generated files (and specify that the extraction rule generates all of them)

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+
+# TODO: relying on directory modification time is a bad idea!
+# Silence `ctrtool`, which is VERY verbose by default
+${vc_dir}/%/: ${vc_dir}/%.orig.cia ${vc_dir}/seeddb.bin
```
Not sure why the wildcard directory target is not working for me... probably more reason to move away from directory targets.

``` Considering target file 'poke-cia/pokegold/'.
     File 'poke-cia/pokegold/' does not exist.
     Finished prerequisites of target file 'poke-cia/pokegold/'.
    Must remake target 'poke-cia/pokegold/'.
make: *** No rule to make target 'poke-cia/pokegold/', needed by 'poke-cia/pokegold/romfs/pokegold.patch'.  Stop.
```

vulcandth replied:
Once I created the directories once manually.. the error went away... Even if I deleted them later... super weird.

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+
+define make_cxi_rule
+$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1)
+	makerom -f cxi -o $$@ -rsf $$< \
+	        -exheader $${vc_dir}/$(1)/exheader.bin \
+	        -logo $${vc_dir}/$(1)/logo.lz \
+	        -plainrgn $${vc_dir}/$(1)/plain.bin \
+	        -code $${vc_dir}/$(1)/exefs/code.bin \
+	        -icon $${vc_dir}/$(1)/exefs/icon.bin \
+	        -banner $${vc_dir}/$(1)/exefs/banner.bin
```
```suggestion
	(cd ${vc_dir}/$(1); \
	makerom -f cxi -o ../$(1).game.cxi -rsf ../game.rsf \
	        -exheader exheader.bin \
	        -logo logo.lz \
	        -plainrgn plain.bin \
	        -code exefs/code.bin \
	        -icon exefs/icon.bin \
	        -banner exefs/banner.bin \
	)
```
It appears makerom needs to run in context of the roms directory. This is why we are getting the ROMFS and MANUAL errors.

jakcron replied:
This is a strange bug with makerom. I'll have to test this.

jakcron replied:
The only thing I can think of is you aren't putting the whole path in quotes. If there is a space in `${vc_dir}/$(1)` it'll probably fail.

jakcron replied:
Quotes as in: `-banner "$${vc_dir}/$(1)/exefs/banner.bin"`

ISSOtm replied:
Yeah, that can in theory be a problem, but it would already break the rules: Make would expand the deps line into `some_dir/a dir with spaces/pokered.game.cxi: [...]`, which would fail beforehand.

And I don't even have any spaces in the absolute path to that directory either (`/home/issotm/pokered`), so I'm quasi-certain it's not that.

jakcron replied:
Well thanks for letting me know then.

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+endef
+$(foreach vc,${vc_name},$(eval $(call make_cxi_rule,${vc})))
+
+${vc_dir}/%.manual.cfa: ${vc_dir}/manual.rsf
+	makerom -f cfa -o $@ -rsf $<
```
Same as the above suggestion; This also needs to run in the context of the roms directory. Maybe we can use another rule here?

vulcandth replied:
Same as the `makerom -f cxi` this needs to be run from the context of inside the roms directory.

```suggestion
define make_cfa_rule
$${vc_dir}/$(1).manual.cfa: $${vc_dir}/manual.rsf
	(cd ${vc_dir}/$(1); \
	makerom -f cfa -o ../$(1).manual.cfa -rsf ../game.rsf \
	)
endef
$(foreach vc,${vc_name},$(eval $(call make_cfa_rule,${vc})))
```

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+vc_orig_cia   := $(addsuffix .orig.cia, ${vc_rom_dirs})
+vc_game_cxi   := $(addsuffix .game.cxi, ${vc_rom_dirs})
+vc_manual_cfa := $(addsuffix .manual.cfa, ${vc_rom_dirs})
+
```
This should probably be included, in order to specify the names of the requirements of cxi

```suggestion
vc_cxi_req    := exheader.bin logo.lz plain.bin exefs/banner.bin exefs/code.bin exefs/icon.bin romfs/

```

ISSOtm replied:
Yeah, it'll be part of getting rid of the directory dependency if we go with the second option I suggested above—so I'll take this as your agreement that we should go with that ;3

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+$(foreach vc,${vc_name},$(eval $(call copy_rom_rule,${vc})))
+
+define make_cxi_rule
+$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1)
```
To go with the `vc_cxi_req` I suggested earlier.

```suggestion
$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1) $(addprefix $${vc_dir}/$(1)/, $(vc_cxi_req))
```

This command should be re-run when any of the `vc_cxi_req` files are updated. The only issue is i'm not sure how to re-run `${vc_dir}/%/:` if one of these files are deleted, or missing.

ISSOtm replied:
The fix is to simply make the "directory rule" have these files as targets instead.

vulcandth commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
+${vc_dir}/%.manual.cfa: ${vc_dir}/manual.rsf
+	makerom -f cfa -o $@ -rsf $<
+
+${vc_dir}/%.cia: ${vc_dir}/%.game.cxi ${vc_dir}/%.manual.cfa
+	makerom -f cia -o $@ -content $<:0:0 -content ${vc_dir}/$*.manual.cfa:1:1
```
Would this be better if it were like this:

```suggestion
define make_cia_rule
$${vc_dir}/$(1).cia: ${vc_dir}/$(1).game.cxi ${vc_dir}/$(1).manual.cfa
	makerom -f cia -o $$@ -content $$<:0:0 -content ${vc_dir}/$(1).manual.cfa:1:1
endef
$(foreach vc,${vc_name},$(eval $(call make_cia_rule,${vc})))
```

Just curious

ISSOtm replied:
Yes and no. Implicit rules are IMO easier to read, but they also generate less user-friendly errors when their prereqs are missing (you'll get "no rules for X" instead of "missing prereq Y for X", due to their nature).

vulcandth commented:
What I think the PR needs left to do: (let me know what else you think of)

*   [x] Remove reliance on the "extract" directories' timestamp.
*   [x] Add a rule to only extract to allow the user to view / manually update things, before repackaging.
*   [x] Verify the dependencies for each rule are correct, and will execute appropriately when things are changed. (Mostly just testing on my part)
*   [x] Add configurable option in the config to enable/disable ctrtool output suppression. Make it easier to turn back on the noisy output for troubleshooting.

jakcron commented on cia.mk:
```diff
@@ -1,59 +1,91 @@
-$(vc_cia): $(vc_game_cxi) $(vc_manual_cfa)
-	makerom -f cia -o $@ -content $(basename $@).game.cxi:0:0 -content $(basename $@).manual.cfa:1:1
\ No newline at end of file
+	ctrtool --contents=$@contents $< >/dev/null
```
if you are using a CTRTool v1.0.0 and newer, you don't have to extract the CIA to process the contents.

Instead of doing this
```
	ctrtool --contents=$@contents $< >/dev/null
	ctrtool --seeddb=${vc_dir}/seeddb.bin \
 	        --exheader=$@exheader.bin \
 	        --exefsdir=$@exefs \
 	        --romfsdir=$@romfs \
 	        --logo=$@logo.lz \
 	        --plainrgn=$@plain.bin \
 	        $@contents.0000.* >/dev/null
	ctrtool --seeddb=${vc_dir}/seeddb.bin \
 	        --romfsdir=$@manual \
 	        $@contents.0001.* >/dev/null
```
Try:
```
	ctrtool --cidx 0 \
 	        --seeddb=${vc_dir}/seeddb.bin \
 	        --exheader=$@exheader.bin \
 	        --exefsdir=$@exefs \
 	        --romfsdir=$@romfs \
 	        --logo=$@logo.lz \
 	        --plainrgn=$@plain.bin \
 	        $< >/dev/null
	ctrtool --cidx 1 \
 	        --seeddb=${vc_dir}/seeddb.bin \
 	        --romfsdir=$@manual \
 	        $< >/dev/null
```

vulcandth replied:
Thanks for the tip! That worked for me after I updated my ctrtool XD.

ISSOtm replied:
Great! I'm using 1.0.1, so I'll try that. We'll have to note that minimum version required—I'll mark 1.0.0, as it's a patch-level difference.

ISSOtm committed:
Allow building out-of-tree

Allows keeping the pret repo pristine, which avoids accidentally committing
changes related to the use of this repo (mainly, the `-include poke-cia/cia.mk` line)

ISSOtm committed:
Avoid intermediate `contents` folders

As suggested by @jakcron — thanks!
https://github.com/vulcandth/poke-cia/pull/1#discussion_r837056460

ISSOtm committed:
Run relevant `makerom` rules in the "extracted" directory

vulcandth commented:
So I just tested the built .cia files based on the current state of the PR.. and when trying to launch them in Citra, Citra crashes. I noticed also that the output .CIA files are larger than the orig.cia files. Which seems wrong, as we delete more things than we add when building.

I normally test using the Citra 3ds emulator, before final tests on real hardware.. as it's easier and quicker to test. 

Not sure immediately what is causing it to crash Citra. Could be a change in ctrtool, could be that I also need to update makerom.  Since I didn't do that when I updated ctrtool. 

I'll dig around today online reading up more on the tools, and seeing if there is something I'm missing. In the meantime, I'd just continue your overall cleanup.

Edit: I may have messed something up on my end during testing now that I think about it, so I might have not been testing with your changes. since I don't remember having to rename my cia-config.mk to config.mk I wonder if I copied the wrong way haha. It was 6:00am in the morning.. so..my brain wasn't functioning properly.

Edit Edit: I definitely was not using your changes.... I know what I did wrong in that aspect.. I don't feel like admitting the problem... It was early.. my brain wasn't working. Doesn't mean there isn't a problem with the .cia files... but I know at least it wasn't a good test from what you've been doing.

ISSOtm commented:
If you still get the crashes, test with the first commit before my PR, and if works there, bisect?

vulcandth commented:
> If you still get the crashes, test with the first commit before my PR, and if works there, bisect?

When I get home i'll build it manually and verify it works.. then i'll step through *all* of the commits until it stops working. Hopefully it was just something I messed up this morning. I was rushing through the test.

ISSOtm commented:
Hey, don't rush yourself! :P I'm not gonna go away

ISSOtm committed:
Rename "vc_name" to "roms_names"

More descriptive, and conveys that there may be multiple names better

ISSOtm committed:
Get rid of `vc_` prefix

ISSOtm committed:
Avoid relying on directory modification time

ISSOtm committed:
Add option to allow toggling `ctrtool` verbosity from the CLI

ISSOtm committed:
Rebuild CXA file if contents change

Caveat: deleting files in the `romfs` folder will not trigger an update;
properly handling this is much more involved, and requires generating dep files.

ISSOtm commented:
That last commit fixes your second bullet point, and the next one will implement the second one. As mentioned, it's not entirely fool-proof, as deleting files in the `romfs` folder will not trigger a rebuild. I think it should be fine for this use case, however.

ISSOtm committed:
Add "extract" target

ISSOtm committed:
Do not delete `seeddb.bin` with `distclean`

It's not a generated file

ISSOtm commented:
Oh yeah, the `config.mk.template` file was missing. lol

ISSOtm committed:
Allow passing `ctrtool` and `makerom` paths on the command line

ISSOtm committed:
Document required version of disassemblies more precisely

ISSOtm committed:
Clean up installation / usage instructions a bit

ISSOtm commented:
And with that... I am personally happy with the state of the repo! All that remains is a new review, and testing.

vulcandth commented on README.md:
```diff
@@ -1,47 +1,70 @@
+  (Relative paths must be relative to the `poke-cia` directory.)
 
- `poke-cia/pokecrystal11.orig.cia`
+Copy your original dumped .cia files to `<build_name>.orig.cia`, where `<build_name>` is one of the names you put in `roms_names`.
```
```suggestion
Copy and rename your original dumped .cia files to `<build_name>.orig.cia`, where `<build_name>` is one of the names you put in `roms_names`.
```

vulcandth commented on README.md:
```diff
@@ -1,47 +1,70 @@
+# CIA building extension for Pret Pokemon Gen I / II Repos
 
-This repo provides a simple extension that integrates the build of a Virtual Console .cia file to the pret pokemon Gen I / II repos, to ease the building of a VC .cia for your ROM Hack down to something as simple as `make cia`.
+This repo provides a simple extension that integrates the build of a Virtual Console .cia file to the pret pokemon Gen I / II repos, to ease the building of a VC .cia for your ROM Hack down to something as simple as typing `make`.
```
```suggestion
This repo provides a simple extension that repackages a Nintendo 3DS Virtual Console (VC) `.cia` file using the built `.gbc`(s) and `.patch`(s) generated from the Pret Pokémon Gen I / II repos. This will ease the building of a VC `.cia` for your ROM Hack down to something as simple as typing `make`.
```

vulcandth commented on config.mk.template:
```diff
@@ -0,0 +1,11 @@
+# Uncomment one of the following lines depending on your repo, or write your own!
+#roms_names := pokered pokeblue
+#roms_names := pokeyellow
+#roms_names := pokegold pokesilver
+#roms_names := pokecrystal11
+#roms_names := my_awesome_hack
```
Is the double plural `roms_names` a personal preference?

ISSOtm replied:
I think `rom_names` is the correct form, but I don't like that `rom` is singular nonetheless. I'll let you pick whichever you think is best.

vulcandth commented on README.md:
```diff
@@ -1,47 +1,70 @@
-CIA building extension for Pret Pokemon Gen I / II Repos
-======================================
+# CIA building extension for Pret Pokemon Gen I / II Repos
```
```suggestion
# CIA building extension for Pret Pokémon Gen I / II Repos
```

vulcandth commented on Makefile:
```diff
@@ -0,0 +1,125 @@
+# so they are given an order-only dep on the directory (`%/` here).
+# Do NOT depend on it directly, as directory modification times update in unintuitive ways!
+#
+# This extracts the original CIA's contents, but deletes the original ROM and patch
```
```suggestion
# This extracts the original CIA's contents, deletes the original ROM, and deletes all patch files. 
# (There are extra, un-needed patch files left in by the VC developers for the other games.)
```

vulcandth commented on Makefile:
```diff
@@ -0,0 +1,125 @@
+## "Option" variables, intended to be overridden from the command line
+# (Or from `config.mk` if you want to persist those)
+
+# Clear this to let ctrtool speak its mind. It is quite talkative.
+VERBOSE_CTRTOOL := >/dev/null
+
+# Paths to the executables (mind that lack of slashes means PATH will be searched instead!)
+CTRTOOL := ctrtool
+MAKEROM := makerom
```
These should be moved to the config file. If users want to modify them, they shouldn't have to deal with git complaining about differences.

ISSOtm replied:
Not quite, they can be overridden in the config file without conflicts. (Writing `CTRTOOL := path/to/ctrtool` inside of it will override this default value.)

vulcandth reviewed (changes requested):
Overall an amazing set of changes! You have been incredibly helpful. Most of what I have is just documentation wording changes i'd like included in the PR. After this, all that is left is for me to do is a series of comprehensive testing to verify it builds properly for all the repos, and to verify that the .cia files work properly in Citra and Real Hardware. If that all checks out, we can merge. I'll start on this when I get home from work!

vulcandth committed:
Update Readme and Comments From Review.

vulcandth committed:
Makerom Directory context is for cxi and cfa only.

Missing directory context for makerom manual.cfa
Do no need directory context for makerom cia output

vulcandth committed:
roms_names -> rom_names

vulcandth commented:
The .CIA files fail to run on Citra and/or real hardware when using ctr-tool newer than `0.7` . I will open an issue on the ctr-tool github.

vulcandth commented:
@ISSOtm As soon as the issue with failing to run on Citra/Real Hardware is resolved; We can merge this PR. Everything else looks great. I'd rather not have to enforce ctrtool version 0.7.. and go back to the intermediate extraction process. 

(I also tested the intermediate extraction process with the newer version.. and although it *appears* to build successfully as well, it also fails to run on citra/real hardware)

vulcandth committed:
Readme version update and Special Credits.
